File: examples/typescript/good/isp.ts
interface Attakable {
  attack(other: string): void;
}

interface Talkable {
  talk(other: string): void;
}

interface Movable {
  move(x: number, y: number): void;
}

class GoodMonster implements Attakable, Movable {
  attack(other: string): void {
    console.log(`Monster attacks ${other}`);
  }

  move(x: number, y: number): void {
    console.log(`Monster moves to (${x}, ${y})`);
  }
}

class GoodNPC implements Talkable {
  talk(other: string): void {
    console.log(`NPC talk to ${other}`);
  }
}

(function () {
  const monster = new GoodMonster();
  const npc = new GoodNPC();

  monster.move(10, 15);
  monster.attack("john");
  npc.talk("john");
})();

File: examples/python/good/isp.py
import abc

class Attackable(abc.ABC):
    def attack(self, other):
        print(f"I attack {other}")

class Talkable(abc.ABC):
    def talk(self, other):
        print(f"I talk to {other}")

class Movable(abc.ABC):
    def move(self, x, y):
        print(f"I move to ({x}, {y})")

class NPC(Talkable):
    def talk(self, other):
        print(f"NPC talk to {other}")

class Monster(Attackable, Movable):
    def attack(self, other):
        print(f"Monster attack {other}")
  
    def move(self, x, y):
        print(f"Monster move to ({x}, {y})")


monster = Monster()
monster.move(10, 15)
monster.attack("john")
npc = NPC()
npc.talk("jane")